pandas/core/generic.py

```diff
@@ -5906,6 +5908,129 @@ def infer_objects(self: FrameOrSeries) -> FrameOrSeries:
+            new_dtype = dtype
+            try:
+                result = self.astype(target_int_dtype)
+                new_dtype = target_int_dtype
+                changeit = False
+                constructit = False
+            except TypeError:
+                pass
+
+        if changeit:
+            if new_dtype == "integer":
+                new_dtype = {
+                    sd.type: sd.name
+                    for sd in registry.dtypes
+                    if isinstance(sd.name, str) and "Int" in sd.name
+                }.get(dtype.type, target_int_dtype)
+            result = self.astype(new_dtype)
+        else:
+            if constructit:
+                result = self._constructor(self).__finalize__(self)
+
+        return result
+
+    def as_nullable_dtypes(self: FrameOrSeries) -> FrameOrSeries:
+        """
+        Convert columns of DataFrame or a Series to types supporting ``pd.NA``.
+
+        | If the dtype is "object", convert to "string", "boolean" or an appropriate
+          integer type.
+        | If the dtype is "integer", convert to an appropriate integer type.
+        | If the dtype is numeric, and consists of all integers, convert to an
+          appropriate type.
+
+        Returns
+        -------
+        converted : a copy of the same type as caller
+
+        Examples
+        --------
+        >>> df = pd.DataFrame(
+        ...     {
+        ...         "a": pd.Series([1, 2, 3], dtype=np.dtype("int32")),
+        ...         "b": pd.Series(["x", "y", "z"], dtype=np.dtype("O")),
+        ...         "c": pd.Series([True, False, np.nan], dtype=np.dtype("O")),
+        ...         "d": pd.Series(["h", "i", np.nan], dtype=np.dtype("O")),
+        ...         "e": pd.Series([10, np.nan, 20], dtype=np.dtype("float")),
+        ...         "f": pd.Series([np.nan, 100.5, 200], dtype=np.dtype("float")),
+        ...     }
+        ... )
+
+        >>> df
+           a  b      c    d     e      f
+        0  1  x   True    h  10.0    NaN
+        1  2  y  False    i   NaN  100.5
+        2  3  z    NaN  NaN  20.0  200.0
+
+        >>> df.dtypes
+        a      int32
+        b     object
+        c     object
+        d     object
+        e    float64
+        f    float64
+        dtype: object
+
+        >>> dfn = df.as_nullable_dtypes()
+        >>> dfn
+           a  b      c     d     e      f
+        0  1  x   True     h    10    NaN
+        1  2  y  False     i  <NA>  100.5
+        2  3  z   <NA>  <NA>    20  200.0
+
+        >>> dfn.dtypes
+        a      Int32
+        b     string
+        c    boolean
+        d     string
+        e      Int64
+        f    float64
+        dtype: object
+        """
+        if self.ndim == 1:
+            return self._as_nullable_dtype()
+        else:
+            results = [col._as_nullable_dtype() for col_name, col in self.items()]
+            result = pd.concat(results, axis=1, copy=False)
+            result.columns = self.columns
```

Comment:
Why do you need this assignment?

Reply:
Following pattern in `DataFrame.astype()`.  I'll try without.

Reply:
works without.